comprl/server/interfaces.py

```diff
@@ -58,109 +70,207 @@ def disconnect(self, reason: str):
+        players (dict[PlayerID, IPlayer]):
+            A dictionary of players participating in the game.
+        start_time (datetime):
+            The start time of the game.
+        finish_callbacks (list[Callable[["IGame"], None]]):
+            A list of callbacks to be executed when the game ends.
+    """
 
+    @abc.abstractmethod
     def __init__(self, players: list[IPlayer]) -> None:
-        self.players: list[IPlayer] = players
-        self.current_actions: list = [None for _ in players]
-        self.result_received: int = 0
-        self.id: GameID = id_generator.generate_game_id()
-        self.start_time = None
+        """
+        Initializes a new instance of the IGame class.
+
+        Args:
+            players (list[IPlayer]): A list of players participating in the game.
+        """
+        self.id: GameID = IDGenerator.generate_game_id()
+        self.players = {p.id: p for p in players}
+        self.start_time = datetime.now()
+        self.finish_callbacks: list[Callable[["IGame"], None]] = []
 
-        self.finish_callbacks: list[Callable] = []
+    def add_finish_callback(self, callback: Callable[["IGame"], None]) -> None:
+        """
+        Adds a callback function to be executed when the game ends.
 
-    def add_finish_callback(self, callback: Callable) -> None:
-        """link a callback to the end of a game"""
+        Args:
+            callback (Callable[["IGame"], None]): The callback function to be added.
+        """
         self.finish_callbacks.append(callback)
 
     def start(self):
         """
-        notifies all players that the game has started
-        and starts the game cycle
+        Notifies all players that the game has started and starts the game cycle.
         """
-
         self.start_time = datetime.now()
 
-        for p in self.players:
-            p.notify_start(game_id=self.id)
+        for player in self.players.values():
+            player.notify_start(self.id)
 
-        self._game_cycle()
+        self._run()
 
     def end(self, reason="unknown"):
-        """notifies all players that the game has ended
+        """
+        Notifies all players that the game has ended.
 
         Args:
-            reason (str, optional): reason why the game has ended. Defaults to "unknown"
+            reason (str): The reason why the game has ended. Defaults to "unknown".
         """
-        for c in self.finish_callbacks:
-            c(self.id)
 
-        # we might want to move this
-        for i, p in enumerate(self.players):
-            p.notify_end(result=self._player_won(i), stats=self._player_stats(i))
+        for callback in self.finish_callbacks:
+            callback(self)
 
-    @abc.abstractmethod
-    def _update_environment(self):
-        """works with the current_actions list to change the environment accordingly."""
-        ...
+        for player in self.players.values():
+            player.notify_end(False, 1)
```

Comment:
Still uses dummy values `False` and `1`

Reply:
Well there is not really a function defined for the stats thats why...

Reply:
There are the `get_player_result` and `_player_won`, that should be used for this